linux/devdoc/threadpool_linux_requirements.md

```diff
@@ -0,0 +1,258 @@
+
+If any error occurs, `threadpool_create` shall fail and return `NULL`.
+
+### threadpool_destroy
+
+```C
+MOCKABLE_FUNCTION(, void, threadpool_destroy, THREADPOOL_HANDLE, threadpool);
+```
+
+`threadpool_destroy` frees the resouces associated with `threadpool`.
+
+If `threadpool` is `NULL`, `threadpool_destroy` shall return.
+
+`threadpool_destroy` shall implicit close if `threadpool` state is set to `THREADPOOL_STATE_OPEN`. 
+
+`threadpool_destroy` shall free the resources associated with the threadpool handle. 
+
+`threadpool_destroy` shall destroy the semphore by calling `sem_destroy`.
+
+`threadpool_destroy` shall destroy the SRW lock by calling `srw_lock_destroy`.
+
+### threadpool_open_async
+
+```C
+MOCKABLE_FUNCTION(, int, threadpool_open_async, THREADPOOL_HANDLE, threadpool, ON_THREADPOOL_OPEN_COMPLETE, on_open_complete, void*, on_open_complete_context);
+```
+
+`threadpool_open_async` opens the threadpool asynchronously.
+
+If `threadpool` is `NULL`, `threadpool_open_async` shall fail and return a non-zero value. 
+
+`threadpool_open_async` shall set the state to `THREADPOOL_STATE_OPENING`. 
+
+If `threadpool` has already been opened, `threadpool_open_async` shall fail and return a non-zero value.
+
+`threadpool_open_async` shall create the threads for `threadpool` using `ThreadAPI_Create`.
+
+If one of the thread creation fails, `threadpool_open_async` shall fail and return a non-zero value, terminate all threads already created, indicate an error to the user by calling the `on_open_complete` callback with `THREADPOOL_OPEN_ERROR` and set `threadpool` state to `THREADPOOL_STATE_NOT_OPEN`.
+
+Otherwise, `threadpool_open_async` shall set the state to `THREADPOOL_STATE_OPEN`, indicate open success to the user by calling the `on_open_complete` callback with `THREADPOOL_OPEN_OK` and return zero.
+
+### threadpool_close
+
+```C
+MOCKABLE_FUNCTION(, void, threadpool_close, THREADPOOL_HANDLE, threadpool);
+```
+
+`threadpool_close` closes the threadpool.
+
+If `threadpool` is `NULL`, `threadpool_close` shall fail and return.
+
+Otherwise, `threadpool_close` shall terminate all threads in the `threadpool` and set the state to `THREADPOOL_STATE_NOT_OPEN`.
+
+### threadpool_schedule_work
+
+```C
+MOCKABLE_FUNCTION(, int, threadpool_schedule_work, THREADPOOL_HANDLE, threadpool, THREADPOOL_WORK_FUNCTION, work_function, void*, work_function_context);
+```
+
+`threadpool_schedule_work` schedule the threadpool to work.
+
+If `threadpool` is `NULL`, `threadpool_schedule_work` shall fail and return a non-zero value.
+
+If `work_function` is `NULL`, `threadpool_schedule_work` shall fail and return a non-zero value.
+
+`threadpool_schedule_work` shall get the threadpool state by calling `interlocked_add`.
+
+`threadpool_schedule_work` shall validate that it's state is THREADPOOL_STATE_OPEN and if not shall fail and return a non-zero value.
+
+`threadpool_schedule_work` shall increment the count of pending call that are in progress to be executed.
+
```

Comment:
First we would have to increment the count and THEN check the state.
Because otherwise there is a race condition with the close call.

The following could happen:
i) assume pending call count is 0 before the call to threadpool_schedule_work
ii)threadpool_schedule_work is called (for the thread T0) and checks the state and sees it as THREADPOOL_STATE_OPEN.
iii) in the meanwhile a threadpool_close executes on another thread (T1) and sees the call count as 0, so it changes the state to CLOSING.
iv) thread switch happens back to T0 and now T0 proceeds to increment the call count to 0 and thinks that we're open (but we are NOT!!!). And it proceeds to walk all over data that is also worked on in close. And that's a memory corruption for you right there.

Reply:
Of course since we want to switch to sm, you won't need to take care of this, because sm does that for you.

Reply:
changed to sm_exec_begin